### PR TITLE
[lib] handling nulls

### PR DESCRIPTION
**Pull request by dweepgogia**
- For reporting warnings and errors in case of optional and mandatory fields missing, respectively.

### CODE REVIEW

File: pkg/validate/serialize.go [hunk 1]
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
...
+	for i := 0; i < v.NumField(); i++ {
+
+		fieldValue := v.Field(i)
 
-	return &csv, nil
+		tag := v.Type().Field(i).Tag.Get("json")
+		fields := strings.Split(tag, ",")
+		omitEmptyTagPresent := contains(fields, "omitempty")
 
+		if tag == "" {
```
Comment (gallettilance) [inline pkg/validate/serialize.go:135]: Please explain in a comment why this is needed

File: pkg/validate/serialize.go [hunk 2]
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
...
+	default:
+		return []string{"Error: input file is not aligned with Operator-Lifecycle-Manager's CSV type."}
 	}
+}
+
+// Takes in a string slice and checks if a string (x) is present in the slice.
+// Return true if the string is present in the slicec.
+func contains(a []string, x string) bool {
+	for _, n := range a {
+		if strings.ToLower(x) == strings.ToLower(n) {
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: why use `.ToLower()` here?

File: pkg/validate/serialize.go [hunk 3]
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
...
+
+func hasEmptyField(val interface{}) []string {
+
+	fieldValue := reflect.ValueOf(val)
+
+	switch fieldValue.Kind() {
+	case reflect.Struct:
+		return checkStructHasEmptyFields(fieldValue, []string{})
+	default:
+		return []string{"Error: input file is not aligned with Operator-Lifecycle-Manager's CSV type."}
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: suggestion: `Error: input file is not a valid CSV`

File: pkg/validate/serialize.go [hunk 4]
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
...
+		return csv, err
+	}
+	if err := json.Unmarshal(rawJson, &csv); err != nil {
+		return csv, fmt.Errorf("error parsing CSV list (JSON) : %s", err)
+	}
+
+	return csv, nil
+}
+
+func hasEmptyField(val interface{}) []string {
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: The function name implies you are returning a `bool`. Please change the name to better match its function

File: pkg/validate/serialize.go [hunk 5]
```diff
@@ -17,12 +20,32 @@ func Verify(yamlFileName string) error {
...
 		return fmt.Errorf("Error unmarshalling YAML to OLM's csv type for %s file:  #%s ", yamlFileName, err)
 	}
 
-	fmt.Printf("%s is verified", yamlFileName)
+	// Flag for checking if there's any mandatory field missing.
+	checkError := false
+
+	// Contains error logs for all missing optional and mandatory fields.
+	log := hasEmptyField(csv)
+	sort.SliceStable(log, func(i, j int) bool {
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: Why are you doing this?

File: pkg/validate/serialize.go [hunk 6]
```diff
@@ -31,14 +54,93 @@ func Verify(yamlFileName string) error {
...
+			return true
+		}
+	}
+	return false
+}
+
+// Recursive function that traverses a nested struct passed in as reflect value, and reports for errors/warnings
+// in case of null struct field values.
+// Returns a log of errors as slice of strings.
+func checkStructHasEmptyFields(v reflect.Value, log []string) []string {
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: This function name could be improved. I'd suggest something along the lines of `csvReport` or `csvInspect`

**Review by gallettilance — CHANGES_REQUESTED** [review]
Great work! A couple comments mostly about names. Also please change the commit title to `[lib] - Handling mandatory and optional fields` or something similar and include a space between the title and the description.

File: pkg/validate/serialize.go (hunk 2 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: I want it to ignore case while comparing the strings.

Comment (dweepgogia) [reply pkg/validate/serialize.go]: I could also use `strings.EqualFold` for that if you think that's more appropriate.

File: pkg/validate/serialize.go (hunk 3 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Done

File: pkg/validate/serialize.go (hunk 4 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Done.

File: pkg/validate/serialize.go (hunk 5 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: To print the `Warnings` followed by `Errors` separately.

File: pkg/validate/serialize.go (hunk 6 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: How about I change `hasEmptyField` to `csvInspect` and `checkStructHasEmptyFields` to `getMissingFields` or `getEmptyFields` because `checkStructHasEmptyFields` has nothing to do with `csv` in general?

File: pkg/validate/serialize.go (hunk 2 above)
Comment (gallettilance) [reply pkg/validate/serialize.go]: I understand but is there a situation where `omitempty` has upper case letters?

File: pkg/validate/serialize.go (hunk 5 above)
Comment (gallettilance) [reply pkg/validate/serialize.go]: ah I see. Nice touch but I am not sure if this is needed. We should just display them in the order in which they appear for now. We can always change later based on feedback

File: pkg/validate/serialize.go (hunk 6 above)
Comment (gallettilance) [reply pkg/validate/serialize.go]: Great point! I agree with `csvInspect` instead of `hasEmptyField` and `checkMissingFields` instead of `checkStructHasEmptyFields`

File: pkg/validate/serialize.go (hunk 2 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Not for `omitempty`, but I was thinking of using this helper in some other part of the code. Although, that did not work out, I decided to make it more general so it can be used in other parts.

File: pkg/validate/serialize.go (hunk 5 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Sure.

File: pkg/validate/serialize.go (hunk 6 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Done.

File: pkg/validate/serialize.go (hunk 1 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:135]: Done.

**Comment by dweepgogia** [comment]
Pushed the changes.

File: pkg/validate/serialize.go [hunk 7]
```diff
@@ -17,12 +19,30 @@ func Verify(yamlFileName string) error {
...
 	// Value returned is a marshaled go type (CSV Struct).
-	_, err = unmarshal(rawYaml)
+	csv, err := unmarshal(rawYaml)
 	if err != nil {
 		return fmt.Errorf("Error unmarshalling YAML to OLM's csv type for %s file:  #%s ", yamlFileName, err)
 	}
 
-	fmt.Printf("%s is verified", yamlFileName)
+	// Flag for checking if there's any mandatory field missing.
+	checkError := false
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: Please rename to something more descriptive of the type. For example `isMandatoryFieldMissing` or `isErrorPresent`

File: pkg/validate/serialize.go [hunk 8]
```diff
@@ -31,14 +51,94 @@ func Verify(yamlFileName string) error {
...
 // (2) there is a problem while unmarshalling in go type.
 // Returns an object of type olm.ClusterServiceVersion.
-func unmarshal(rawYAML []byte) (*olm.ClusterServiceVersion, error) {
+func unmarshal(rawYAML []byte) (olm.ClusterServiceVersion, error) {
 
 	var csv olm.ClusterServiceVersion
 
-	if err := yaml.Unmarshal(rawYAML, &csv); err != nil {
-		return nil, fmt.Errorf("error parsing CSV list (JSON) : %s", err)
+	rawJson, err := yaml.YAMLToJSON(rawYAML)
```
Comment (gallettilance) [inline pkg/validate/serialize.go:78]: why did you remove the old way of unmarshalling here?

File: pkg/validate/serialize.go [hunk 9]
```diff
@@ -31,14 +51,94 @@ func Verify(yamlFileName string) error {
...
+	case reflect.Uint, reflect.Uint8, reflect.Uint16, reflect.Uint32, reflect.Uint64, reflect.Uintptr:
+		return v.Uint() == 0
+	case reflect.Float32, reflect.Float64:
+		return v.Float() == 0
+	case reflect.Interface, reflect.Ptr:
+		return v.IsNil()
+	default:
+		panic(fmt.Sprintf("%v kind is not supported.", v.Kind()))
+	}
 }
```
Comment (gallettilance) [inline pkg/validate/serialize.go:211]: Nit: please add a new line at the end of the file

File: pkg/validate/serialize.go [hunk 10]
```diff
@@ -31,14 +51,94 @@ func Verify(yamlFileName string) error {
...
+	return log
+}
+
+// Uses reflect package to check if the value of the object passed is null, returns a boolean accordingly.
+func isEmptyValue(v reflect.Value) bool {
+
+	switch v.Kind() {
+	case reflect.Array, reflect.Map, reflect.Slice, reflect.String:
+		return v.Len() == 0
+	case reflect.Int, reflect.Int8, reflect.Int16, reflect.Int32, reflect.Int64:
```
Comment (gallettilance) [inline pkg/validate/serialize.go:193]: This seems wrong. What if the value is actually 0?

**Review by gallettilance — CHANGES_REQUESTED** [review]

File: pkg/validate/serialize.go (hunk 7 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Done.

File: pkg/validate/serialize.go (hunk 8 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:78]: Because we only have json tags in CSV.

File: pkg/validate/serialize.go (hunk 9 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:211]: Done.

File: pkg/validate/serialize.go (hunk 8 above)
Comment (gallettilance) [reply pkg/validate/serialize.go:78]: right. my bad!

File: pkg/validate/serialize.go (hunk 5 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Reverted back, based on Eric's feedback.

**Comment by dweepgogia** [comment]
Pushed changes. Please review.

File: pkg/validate/serialize.go [hunk 11]
```diff
@@ -17,12 +21,35 @@ func Verify(yamlFileName string) error {
...
+	for i := range log {
+		if strings.Contains(strings.ToLower(log[i]), "error") {
+			isMandatoryFieldMissing = true
+		}
+		fmt.Println(log[i])
+	}
+
+	if isMandatoryFieldMissing {
+		fmt.Printf("Populate all the mandatory fields missing from %s file.", yamlFileName)
+		os.Exit(1)
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: if you are going to exit the function without returning an `error`, might as well remove that from the function definition and have it not return anything.

File: pkg/validate/serialize.go [hunk 12]
```diff
@@ -17,12 +21,35 @@ func Verify(yamlFileName string) error {
...
+	} else {
+		fmt.Printf("%s is verified.", yamlFileName)
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: then here you would just exit(0) I guess

File: pkg/validate/serialize.go [hunk 13]
```diff
@@ -31,14 +58,128 @@ func Verify(yamlFileName string) error {
...
+			} else if emptyVal && !isOptionalField {
+				if parent == "" {
+					log = append(log, fmt.Sprintf("ERROR: Mandatory Struct missing (%s).", newParent))
+				} else {
+					log = append(log, fmt.Sprintf("ERROR: Mandatory Struct missing (%s).", newParent))
+				}
+				continue
+			}
+			log = checkMissingFields(fieldValue, newParent, log)
+		default:
```
Comment (gallettilance) [inline pkg/validate/serialize.go:157]: most of this code is redundant. Maybe you should do the empty check first and then do the switch statement on the kind to see if you have to recursively call the function

**Review by gallettilance — CHANGES_REQUESTED** [review]

File: pkg/validate/serialize.go (hunk 5 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Now that we have separate objects for errors and warnings, we do not need to explicitly sort them. So the order of fields is maintained and they are segregated as well :)

File: pkg/validate/serialize.go (hunk 11 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Not making this change in the CLI in this PR.

**Comment by dweepgogia** [comment]
Pushed the changes. Please review.

File: pkg/validate/serialize.go [hunk 14]
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
...
 	if err != nil {
 		return fmt.Errorf("Error unmarshalling YAML to OLM's csv type for %s file:  #%s ", yamlFileName, err)
 	}
 
-	fmt.Printf("%s is verified", yamlFileName)
-	return nil
+	// Contains error logs for all missing optional and mandatory fields.
+	errorLog := csvInspect(csv)
+	checkManifestResult(errorLog.warnings)
+	checkManifestResult(errorLog.errors)
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: can we put this in the if statement below?

File: pkg/validate/serialize.go [hunk 15]
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
...
+	if errorLog.errors != nil {
+		return fmt.Errorf("Populate all the mandatory fields missing from %s file.", yamlFileName)
+	} else {
+		fmt.Printf("%s is verified.", yamlFileName)
+		return nil
+	}
+}
+
+// Iterates over the list of warnings and errors.
+func checkManifestResult(err []missingTypeError) {
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: not sure this is the right name for what this is doing

File: pkg/validate/serialize.go [hunk 16]
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
...
+	if err != nil {
+		for _, v := range err {
+			assertTypeToGetValue(v)
+		}
+		fmt.Println()
+	}
+}
+
+// Asserts type to get the underlying field value.
+func assertTypeToGetValue(v interface{}) {
```
Comment (gallettilance) [inline pkg/validate/serialize.go:63]: this name is strange too

File: pkg/validate/serialize.go [hunk 17]
```diff
@@ -17,28 +33,184 @@ func Verify(yamlFileName string) error {
...
+	return log
+}
+
+// Returns updated error log with missing optional/mandatory field/struct objects.
+func updateLog(log manifestResult, typeName string, newParentStructName string, emptyVal bool, isOptionalField bool) manifestResult {
+
+	if emptyVal && isOptionalField {
+		errString := fmt.Sprintf("Warning: Optional %s Missing.", typeName)
+		log.warnings = append(log.warnings, missingTypeError{errString, typeName, newParentStructName, false})
+		//log = append(log, fmt.Sprintf("WARNING: Optional %s missing (%s).", typeName, newParentStructName))
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: Please remove commented out code

File: pkg/validate/serialize.go (hunk 9 above)
Comment (gallettilance) [reply pkg/validate/serialize.go:211]: did you remove the new line?

**Review by gallettilance — COMMENTED** [review]
Looks good just a few minor things

File: pkg/validate/serialize.go (hunk 17 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Done.

File: pkg/validate/serialize.go (hunk 16 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:63]: Well, the only thing this function is doing is asserting type and getting the value.

File: pkg/validate/serialize.go (hunk 15 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: How about `getErrorsFromManifestResult`?

File: pkg/validate/serialize.go (hunk 14 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Do you want to check for `nil` before calling `checkManifestResult`?

File: pkg/validate/serialize.go (hunk 13 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:157]: Done.

File: pkg/validate/serialize.go (hunk 10 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:193]: Addressed this in a comment.

File: pkg/validate/serialize.go (hunk 9 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:211]: I still have it on my local. Btw what do we need it for?

File: pkg/validate/serialize.go (hunk 14 above)
Comment (gallettilance) [reply pkg/validate/serialize.go]: I just think since you are printing the errors only if it's not nil, and you are raising an error only if its not nil, might as well include all of that under the same IF no?

File: pkg/validate/serialize.go (hunk 15 above)
Comment (gallettilance) [reply pkg/validate/serialize.go]: yeah!

File: pkg/validate/serialize.go (hunk 16 above)
Comment (gallettilance) [reply pkg/validate/serialize.go:63]: But the getting is more of a print no?

File: pkg/validate/serialize.go (hunk 15 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Done.

File: pkg/validate/serialize.go (hunk 16 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go:63]: The way we describe the if statement in line `67` is it is type checking and getting the underlying struct value. See https://golangbot.com/interfaces-part-1/#typeassertion.

Comment (gallettilance) [reply pkg/validate/serialize.go:63]: ah ok gotcha thanks for clarifying!

File: pkg/validate/serialize.go (hunk 14 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: Done.

**Comment by dweepgogia** [comment]
Pushed the changes. Please review.

File: pkg/validate/serialize.go [hunk 18]
```diff
@@ -17,28 +33,181 @@ func Verify(yamlFileName string) error {
...
+		return fmt.Errorf("Populate all the mandatory fields missing from %s file.", yamlFileName)
+	}
+	fmt.Printf("%s is verified.", yamlFileName)
 	return nil
 }
 
+// Iterates over the list of warnings and errors.
+func getErrorsFromManifestResult(err []missingTypeError) {
+	if err != nil {
+		for _, v := range err {
```
Comment (gallettilance) [inline pkg/validate/serialize.go]: what happens if err is nil here?

File: pkg/validate/serialize.go (hunk 9 above)
Comment (gallettilance) [reply pkg/validate/serialize.go:211]: ok ok it's just hard to tell in this UI

File: pkg/validate/serialize.go (hunk 18 above)
Comment (dweepgogia) [reply pkg/validate/serialize.go]: No error/warning gets printed. Control goes back and you just see `csv.yaml file is verified.`

Comment (dweepgogia) [reply pkg/validate/serialize.go]: Do I need to explicitly `return` at the end if err is nil?

**Review by gallettilance — APPROVED** [review]
